### ISSUE TITLE
Nullable state not properly set on variables in pattern on erroneous call

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#67880]
[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EAWXAFQFMBnAFxQAoBKcysjSoNwB6YbgDGEADZTcAdyisAFgHsArq1yMATtpXbmvIbgBuEbfNwBeXAFlOXAIQA6AGrS1jZwElWjMCgA3EZCcs70KgDKrNqwAObcgSJiMCry5jDxIYLZlFAAZrgc9ty4UMxFZhYAHgA0uAD6XDwCxqS5xtXhUTHxicm4qenamTBxHbgAviHTrRQhOPh0TGxI3CH8bQOSMvKKSlq6+oZzQlW4AJ7WdpwxnlxuHl6+/kETF93RsWP9ooNpcgyWQwEwKRRKd0YXDKFQ45wAXvUmi0tpstpR4Z9ej8uEk/oCRsCtrNBCTKAs8BwECgAAz1ak0rgAfhupSsAD5cMo9HJBmoZI5ghhZkA)

```cs
public class C {
    public void Test1()
    {
        // call without errors
        var w = M1()!.Value.Item1;
        w.ToString(); // no warning
        
        if (M1() is (var x, _))
        {
            x.ToString(); // no warning
        }
    }
    
    public void Test2()
    {
        // call with errors
        var y = M1(true).Value.Item1;
        y.ToString(); // no warning

        if (M1(true) is (var z, _))
        {
            z.ToString(); // warning
        }
    }
    
    public (string, string)? M1() => throw null!;
}
```

Expected behavior: No nullable warnings
Actual behavior: Nullable warning on usage of pattern variable, only when call to M1 has errors

**Comment by Rayuga35** [comment]
For Test 2 function: - M(true) throws a null reference exception, the pattern matching is expression in Test2 will never match, and the code inside the if block will not be executed.
To deal with this modification of method M1 to return a non-null value.  For this you we can return tuple of this non-null value
Statement code: - public (string, string) M1() => ("hello", "world");

Suppose if method dont return non-null value then modify the code that calls the method to handle null values properly. For example, you can use the null-coalescing operator  '??'  to provide a default value in case the method returns null:
Statement code:- var result = M1() ?? ("default", "value");

**Comment by Rayuga35** [comment]
public class C 
{
    public void Test1()
    {
        // call without errors
        var w = M1().Value.Item1;
        w.ToString(); // no warning

        if (M1() is (var x, _))
        {
            x.ToString(); // no warning
        }
    }
    
    public void Test2()
    {
        // call without errors
        var y = M1(true).Value.Item1;
        y.ToString(); // no warning

        if (M1(true) is (var z, _))
        {
            z.ToString(); // no warning
        }
    }
    
    public (string, string) M1(bool throwException = false) 
    {
        if (throwException) 
        {
            throw new Exception("M1 method exception");
        }
        
        return ("hello", "world");
    }
}

**Comment by Rayuga35** [comment]
``public class C 
{
    public void Test1()
    {
        // call without errors
        var w = M1().Value.Item1;
        w.ToString(); // no warning

        if (M1() is (var x, _))
        {
            x.ToString(); // no warning
        }
    }
    
    public void Test2()
    {
        // call without errors
        var y = M1(true).Value.Item1;
        y.ToString(); // no warning

        if (M1(true) is (var z, _))
        {
            z.ToString(); // no warning
        }
    }
    
    public (string, string) M1(bool throwException = false) 
    {
        if (throwException) 
        {
            throw new Exception("M1 method exception");
        }
        
        return ("hello", "world");
    }
`}`

**Comment by sdelarosbil** [comment]
Hey, can I take this one?

**Comment by sdelarosbil** [comment]
Huh, this issue just got interesting: the actual error recovery part isn't why this is failing.

It has to do with the BoundIsPatternExpression being wrapped in a BoundConversion to bool. It just happens to repro here because when the expression fails to bind, the compiler recovers by wrapping it in a fake BoundConversion.

But I found it doesn't have to be specific to this: even this code repro the issue without errors:

```cs
#nullable enable
public class C
{
    public (string, string)? M1() => throw null!;

    public void M2()
    {
        if ((bool)(M1() is (var z, _)))
        {
            z.ToString(); // warning
        }
    }
}
```

I have yet to determine the details of why this messes up the nullable analysis (working on it still), but one thing I know for sure: it does indeed correctly determine that z can't be null, but the BoundConversion seems to undo that nullable state

**Comment by sdelarosbil** [comment]
hmmm @RikkiGibson I might need some advice here because I think I understand the issue, but not exactly the proper way to address it.

Basically, in NullableWalker.VisitConversion (the override one), the operand of the conversion is visited via `VisitRvalueWithState` which eventually gets to `VisitRvalueEpilogue` causing an `Unsplit` call.

That Unsplit is what breaks things: right after visiting the BoundIsPatternExpression, the conditional state is setup with a true/false state being correct, but Unsplit will join them thinking they converge when they can't converge. This is because this conversion really can't change the state here since it's an identity conversion (or NoConversion in the case of the erroneous code) over a boolean expression that already had to be boolean because it's a BoundIsPatternExpression.

It honestly seems like that conversion shouldn't even be involved in the analysis because unless I am missing something, I don't think a wrapped BoundIsPatternExpression in a boolean BoundConversion could possibly influence nullability.

If that is the case, how should this be addressed? I can try to bypass the unsplit and just visit the operand node, but I don't know if it's the proper solution here or if I am missing something.

**Comment by RikkiGibson** [comment]
At a glance it feels reasonable to me to replace `TypeWithState operandType = VisitRvalueWithState(operand);` with `Visit(operand); TypeWithState operandType = RvalueResultType;`. However, the behavior of `trackMembers` also needs to be rationalized with this change. I think this flag requires an unsplit state since it is copying member state from the operand to the conversion result. It's possible we could just choose not to do this for boolean expressions, since it doesn't really make sense to track member state for booleans, and we could also unsplit the operand visit result when its type is not boolean to satisfy a few edge case tests. I would say give it a try and see how things shake out.
